### DISCUSSION TITLE
Where should i type PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0  to disable the upper limit

### DISCUSSION THREAD

**Discussion by GondorGitaros** [lllyasviel/Fooocus#1156] [Q&A] [answered]
I am a total noob and I don't know where to type PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0 and i get this error:

RuntimeError: MPS backend out of memory (MPS allocated: 8.89 GB, other allocations: 172.98 MB, max allowed: 9.07 GB). Tried to allocate 6.25 MB on private pool. Use PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0 to disable upper limit for memory allocations (may cause system failure).
Total time: 129.38 seconds

**Comment by mel870** [comment, marked as answer]
Edit the launch.py and add: os.environ["PYTORCH_MPS_HIGH_WATERMARK_RATIO"] = "0.0"

![Screenshot 2023-12-03 at 4 29 44 PM](https://github.com/lllyasviel/Fooocus/assets/7969647/6399799b-f58d-4cc7-9151-b466a98ba198)

**Reply by briandamkroger21** [reply]
Sorry to be dense, but I still don't understand where to type the following:  
          PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0     
in my   launch.py   file.  

When I open the file using TextEdit I see the following:
***********************************************************************
from modules import launch_utils

args = launch_utils.args
python = launch_utils.python
git = launch_utils.git
index_url = launch_utils.index_url
dir_repos = launch_utils.dir_repos

commit_hash = launch_utils.commit_hash
git_tag = launch_utils.git_tag

run = launch_utils.run
is_installed = launch_utils.is_installed
repo_dir = launch_utils.repo_dir

run_pip = launch_utils.run_pip
check_run_python = launch_utils.check_run_python
git_clone = launch_utils.git_clone
git_pull_recursive = launch_utils.git_pull_recursive
list_extensions = launch_utils.list_extensions
run_extension_installer = launch_utils.run_extension_installer
prepare_environment = launch_utils.prepare_environment
configure_for_tests = launch_utils.configure_for_tests
start = launch_utils.start


def main():
    if args.dump_sysinfo:
        filename = launch_utils.dump_sysinfo()

        print(f"Sysinfo saved as {filename}. Exiting...")

        exit(0)

    launch_utils.startup_timer.record("initial startup")

    with launch_utils.startup_timer.subcategory("prepare environment"):
        if not args.skip_prepare_environment:
            prepare_environment()

    if args.test_server:
        configure_for_tests()

    start()


if __name__ == "__main__":
    main()
********************************************************************
So, where, exactly do I type PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0 

Thanks,
Brian

**Reply by briandamkroger21** [reply]
Or perhaps better yet, how can I edit the other allocations referenced in the error message?

(MPS allocated: 4.36 GB, other allocations: 2.22 GB, max allowed: 6.80 GB)

**Reply by GondorGitaros** [reply]
This is not in Fooocus directory. 
you should do this (if didn't do that already):
`git clone https://github.com/lllyasviel/Fooocus.git`
`cd Fooocus`
`conda env create -f environment.yaml`
`conda activate fooocus`
`pip install -r requirements_versions.txt`
Use your text editor of your choice (for the sake of this discussion I use nano)
`nano launch.py` (if you want to use nano and its not installed `sudo apt install nano`)
Edit it (add `os.environ["PYTORCH_MPS_HIGH_WATERMARK_RATIO"] = "0.0"` to the code)
If you used nano hit `ctrl+x` than `y` than `enter`
If you don't have MAC M2:
    `python entry_with_update.py`
if you have MAC M2:
    `python entry_with_update.py --enable-smart-memory`
and it should work.

Let me know if it helps!!

**Comment by GondorGitaros** [comment]
Thank you it worked!! sorry for the stupid question
